2.0/m_joinpartspam.cpp

```diff
@@ -0,0 +1,231 @@
+/*
+ * InspIRCd -- Internet Relay Chat Daemon
+ *
+ *   Copyright (C) 2016 genius3000 <genius3000@g3k.solutions>
+ *
+ * This file is a module for InspIRCd.  InspIRCd is free software: you can
+ * redistribute it and/or modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation, version 2.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "inspircd.h"
+
+/* $ModDesc: Adds channel mode +V to ban a user after x per y joins and parts/quits (join/part spam) */
+/* $ModAuthor: genius3000 */
+/* $ModAuthorMail: genius3000@g3k.solutions */
+/* $ModDepends: core 2.0 */
+
+class joinpartspamsettings
+{
+ public:
+	unsigned int cycles;
+	unsigned int secs;
+	time_t lastcleanup;
+
+	struct Tracking
+	{
+		time_t reset;
+		unsigned int counter;
+	};
+	std::map<std::string, Tracking> cycler;
+
+	joinpartspamsettings(int c, int s) : cycles(c), secs(s)
+	{
+	}
+
+	/* Called by PostJoin to possibly reset a cycler's Tracking and increment the counter */
+	void addcycle(const std::string& mask)
+	{
+		/* If mask isn't already tracked, set reset time
+		 * If tracked and reset time is up, reset counter and reset time
+		 * Also assume another server denied join and set ban, with a user removing
+		 * the ban if counter >= cycles, reset counter and reset time
+		 */
+		if (cycler.count(mask) != 1)
+			cycler[mask].reset = ServerInstance->Time() + secs;
+		else if (ServerInstance->Time() > cycler[mask].reset || cycler[mask].counter >= cycles)
+		{
+			cycler[mask].counter = 0;
+			cycler[mask].reset = ServerInstance->Time() + secs;
+		}
+
+		++cycler[mask].counter;
+
+		this->cleanup();
+	}
+
+	/* Called by PreJoin to check if a cycler's counter exceeds the set cycles,
+	 * Will first clear a cycler if their reset time is up
+	 */
+	bool zapme(const std::string& mask)
+	{
+		/* Only check reset time and counter if they are already tracked as a cycler */
+		if (cycler.count(mask) == 1)
+		{
+			if (ServerInstance->Time() > cycler[mask].reset)
+				cycler.erase(mask);
+			else if (cycler[mask].counter >= cycles)
+			{
+				cycler.erase(mask);
+				return true;
+			}
+		}
+		return false;
```

Comment:
What you can do here to avoid repeated map lookups is to use the `find()` map method like this:

```c++
std::map<std::string, Tracking>::iterator it = cycler.find(mask);
if (it == cycler.end()) // 'mask' is not in the map
    return false;
Tracking& tracking = it->second;

// Rest of the code
if (ServerInstance->Time() > tracking.reset)
 ...
```